### ISSUE TITLE
`ebft_variant` adds a loss term with no causal shift: it scores each position against its own input token, rewarding copying over next-token prediction

### ISSUE THREAD

**Issue by MakazhanAlpamys** [MakazhanAlpamys/Soup#1230]
Found in a code review; reproduced on `main` at `99d553f5` (CPU; kernel plus the real SFT trainer on `hf-internal-testing/tiny-random-gpt2`).

## What happens

Two hand-built predictors over `labels = [5, 7, 2, 9]`: one that is a perfect causal LM (position t puts all mass on token t+1), and one that copies its input (position t puts all mass on token t).

```python
import torch

from soup_cli.utils.ebft_gdpo import apply_ebft_loss

labels = torch.tensor([[5, 7, 2, 9]])
next_token = torch.full((1, 4, 10), -20.0)
for t in range(3):
    next_token[0, t, labels[0, t + 1]] = 20.0  # position t predicts token t+1: a perfect causal LM
copy_input = torch.full((1, 4, 10), -20.0)
for t in range(4):
    copy_input[0, t, labels[0, t]] = 20.0  # position t repeats its own input token

for name, logits in (("perfect next-token predictor", next_token), ("copies its input", copy_input)):
    print(f"{name:29s}",
          round(apply_ebft_loss(logits, labels, variant="structured", temperature=1.0).item(), 3),
          round(apply_ebft_loss(logits[:, :-1], labels[:, 1:], variant="structured",
                                temperature=1.0).item(), 3))
```

```
perfect next-token predictor  30.576 0.0
copies its input              0.0 40.0
```

The first column is the term as shipped and the second is the same kernel on the causally shifted pair. As shipped, the term is minimised by copying the input and penalises the model TRL is actually training toward.

On the real trainer (`SFTTrainerWrapper.setup()` with `training.ebft_variant: structured`, `ebft_temperature: 1.0`; the hook installed by the same call `train()` makes), one batch in eval mode, `compute_loss` split into TRL's own cross-entropy and the added term:

```
trainer.compute_loss total           = 13.507071
TRL's own CE (unpatched compute_loss) = 6.887587
added EBFT term (total - CE)          = 6.619483
CE(logits[:, t], labels[:, t])        = 6.619483   <- unshifted
CE(logits[:, t], labels[:, t+1])      = 6.887588   <- causal (what TRL trains)
|EBFT term - unshifted CE| = 0.00e+00
|EBFT term - shifted CE|   = 2.68e-01
```

The added term equals the *unshifted* cross-entropy exactly. Through the real `train()` (one step, seed 7), the first logged loss is `6.8892` without `ebft_variant` and `13.5217` with it, so the hook is live on the shipped path.

<details><summary>Script for the real-trainer numbers (run as a file; tiny model from the Hugging Face cache)</summary>

```python
import os
import pathlib
import tempfile

import torch
import torch.nn.functional as F

from soup_cli.config.loader import load_config_from_string
from soup_cli.trainer.sft import SFTTrainerWrapper

HERE = pathlib.Path(__file__).parent
with tempfile.TemporaryDirectory(dir=HERE, ignore_cleanup_errors=True) as tmp:
    os.chdir(tmp)
    cfg = load_config_from_string(
        "base: hf-internal-testing/tiny-random-gpt2\n"
        "task: sft\n"
        "data:\n  train: x.jsonl\n  max_length: 64\n  chat_template: chatml\n"
        "training:\n  epochs: 1\n  batch_size: 2\n  quantization: none\n"
        "  ebft_variant: structured\n  ebft_temperature: 1.0\n"
        "output: ./out\n"
    )
    rows = [{"messages": [{"role": "user", "content": f"Count to {i}."},
                          {"role": "assistant", "content": " ".join(str(k) for k in range(i))}]}
            for i in range(3, 7)]
    wrapper = SFTTrainerWrapper(cfg, device="cpu")
    wrapper.setup({"train": rows})
    trainer = wrapper.trainer
    print("EBFT hook installed after setup():", getattr(trainer, "_soup_ebft_wrapped", False))
    # SFTTrainerWrapper.train() installs the hook with exactly this call (sft.py:2174-2175)
    from soup_cli.utils.ebft_gdpo import attach_ebft_compute_loss
    attach_ebft_compute_loss(trainer, wrapper.config.training)
    print("EBFT hook installed after the train()-time attach:", trainer._soup_ebft_wrapped)
    model = trainer.model
    model.eval()  # no dropout, so every forward below sees the same logits
    batch = next(iter(trainer.get_train_dataloader()))
    batch = {k: v for k, v in batch.items() if torch.is_tensor(v)}
    with torch.no_grad():
        total, outputs = trainer.compute_loss(model, dict(batch), return_outputs=True)
        ce, _ = type(trainer).compute_loss(trainer, model, dict(batch), return_outputs=True)
    lab = batch["labels"]
    logits = outputs.logits.float()
    vocab = logits.shape[-1]
    unshifted = F.cross_entropy(logits.reshape(-1, vocab), lab.reshape(-1), ignore_index=-100)
    shifted = F.cross_entropy(logits[:, :-1].reshape(-1, vocab), lab[:, 1:].reshape(-1),
                              ignore_index=-100)
    ebft_term = (total - ce).item()
    print(f"trainer.compute_loss total           = {total.item():.6f}")
    print(f"TRL's own CE (unpatched compute_loss) = {ce.item():.6f}")
    print(f"added EBFT term (total - CE)          = {ebft_term:.6f}")
    print(f"CE(logits[:, t], labels[:, t])        = {unshifted.item():.6f}   <- unshifted")
    print(f"CE(logits[:, t], labels[:, t+1])      = {shifted.item():.6f}   <- causal (what TRL trains)")
    print(f"|EBFT term - unshifted CE| = {abs(ebft_term - unshifted.item()):.2e}")
    print(f"|EBFT term - shifted CE|   = {abs(ebft_term - shifted.item()):.2e}")

    print("\n=== B2: the real train(), one step, with and without ebft_variant (seed 7) ===")
    for extra in ("", "  ebft_variant: structured\n  ebft_temperature: 1.0\n"):
        cfg2 = load_config_from_string(
            "base: hf-internal-testing/tiny-random-gpt2\n"
            "task: sft\n"
            "data:\n  train: x.jsonl\n  max_length: 64\n  chat_template: chatml\n"
            "training:\n  epochs: 1\n  batch_size: 4\n  gradient_accumulation_steps: 1\n"
            "  quantization: none\n  logging_steps: 1\n  seed: 7\n"
            f"{extra}"
            "output: ./out2\n"
        )
        w2 = SFTTrainerWrapper(cfg2, device="cpu")
        w2.setup({"train": rows})
        w2.train()
        first = [e for e in w2.trainer.state.log_history if "loss" in e][0]["loss"]
        label = "ebft_variant: structured" if extra else "no ebft_variant (control)"
        print(f"  {label:28s} hook={getattr(w2.trainer, '_soup_ebft_wrapped', False)!s:5s} "
              f"first logged loss = {float(first):.4f}")
    os.chdir(HERE)
```

</details>

## Why

`src/soup_cli/utils/ebft_gdpo.py:212-220`: the hook passes `outputs.logits` and `inputs["labels"]` to `apply_ebft_loss` unchanged. The kernel (`ebft_gdpo.py:298-378`) takes `log_softmax(logits / temp)` and gathers at `labels` at the same position (`:373-375`). HF/TRL labels are unshifted; the model shifts inside its own loss, and TRL shifts for its token accuracy (`trl/trainer/sft_trainer.py:1292-1293`: `outputs.logits[..., :-1, :]` against `labels[..., 1:]`). The hook is installed by `SFTTrainerWrapper.train()` (`src/soup_cli/trainer/sft.py:2173-2175`). The closed #682 fixed the same missing shift in the ULD distillation loss.

## Fix path

The mechanical part is to shift inside the kernel or the hook (`logits[:, :-1]` against `labels[:, 1:]`), honour `shift_labels` when the batch provides them (TRL's padding-free path, `sft_trainer.py:1289-1290`), and keep `-100` masking and the `strided` subsampling after the shift.

The shift alone is not enough, which is why this needs a decision first. Once shifted, the term at temperature 1 *is* the model's own token-level cross-entropy: on the two predictors above, the shifted term (0.000 / 40.000) equals `F.cross_entropy(logits[:, :-1], labels[:, 1:])` (0.000 / 40.000). The objective would then become CE counted twice, the same as doubling the learning rate on one loss rather than an energy-based term. Either implement the intended EBFT objective from its reference, or refuse `training.ebft_variant` at config parse until it is, with a message saying the variant is not yet a distinct objective.

## Acceptance

- [ ] Kernel test with the two predictors above: the perfect next-token predictor scores about 0 and the copy-input predictor scores high; the test fails if the shift is removed.
- [ ] Real-trainer test (tiny CPU model): the added term is computed on the causal pairing; if it is defined as a temperature-scaled CE, it matches that quantity over the same mask within 1e-6, and otherwise it matches the documented objective.
- [ ] `strided`: after the shift, only every `stride`-th *target* position contributes.
- [ ] Prompt tokens masked to `-100` (`train_on_responses_only`) stay masked after the shift.
- [ ] A packed / padding-free batch that carries `shift_labels` is handled.
